### ISSUE TITLE
`JsonSerializer.Deserialize<TValue>(JsonNode?)` extensions could be improved

### ISSUE THREAD

**Issue by gregsdennis** [dotnet/runtime#84234]
Reading through the [`JsonSerializer` code](https://github.com/dotnet/runtime/blob/8ffad5201e02ec8657065aabc543135903767535/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Node.cs#L30), these extension methods merely write out the JSON data into text and then deserialize that text.  Wouldn't it be faster to deserialize directly from `JsonNode`?

I do see that efforts have been made to utilize spans and buffer pools, which helps with memory consumption.

Also, the current method does provide some benefits:

- single source of truth - all serialization happens through a single code path
- easier to maintian

## Why do I go through `JsonNode`?

Under various scenarios, I may need to inspect or even manipulate the data before deserializing into my model.  Having the data in the mutable JSON data model provided by `JsonNode` allows me to do this.  My process is:

1. Parse to `JsonNode`
2. Muck about with it
3. Deserialize to my model

Step 3 can then be broken down into:

3a. re-serialize the `JsonNode` to text
3b. deserialize my model from text (which means parsing it again)

I think this is rather wasteful when you _could_ just go directly to the model from `JsonNode`. (My previous JSON library, Manatee.Json, worked by deserializing directly from the JSON model.)

**Comment by eiriktsarpalis** [comment]
The problem is that STJ's serialization and deserialization infrastructure is built on top of `Utf8JsonWriter` and `Utf8JsonReader` which are not abstractable types. `Utf8JsonReader` in particular is a ref struct that can only handle UTF-8 encoded buffers. So this is not merely a matter of inefficient code that could be improved, rather, it is an inefficiency forced by lack of abstraction in the underlying reader/converter infrastructure. To be perfectly honest, I don't think it could be improved without introducing major breaking changes.

> I think this is rather wasteful when you could just go directly to the model from JsonNode. (My previous JSON library, Manatee.Json, worked by deserializing directly from the JSON model.)

Absolutely, and this is generally true for serialization libraries that abstract their reader and writer types. For example Json.NET has a dedicated [`JsonReader` implementation](https://github.com/JamesNK/Newtonsoft.Json/blob/0a2e291c0d9c0c7675d445703e51750363a549ef/Src/Newtonsoft.Json/Linq/JTokenReader.cs) for their DOM type.

**Comment by gregsdennis** [comment]
Yeah, that's definitely something to think about.  And creating extended functionality (such as I could do in a library) would be laborious to maintain, trying to keep it in line with what the text deserialization does.

Is there a reason you can't extract an interface from `Utf8JsonReader` and build the serializer on that?  You could create overloads in `JsonConverter<T>` for the read/write methods instead of changing their signature.

I'm not sure what else might use it in a signature...

**Comment by gregsdennis** [comment]
@eiriktsarpalis I see you've put the milestone on here, but I don't know how you typically handle these kinds of issues.  My question has been answered, so feel free to close if you like.

**Comment by eiriktsarpalis** [comment]
I'd be inclined to leave open if that's ok. It's a real problem and it might be the case that future language innovation might make this scenario possible.
